### PR TITLE
Draft: CCM-7409: add sandbox

### PR DESCRIPTION
**Pull request by harrim91**
## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: .github/workflows/stage-4-acceptance.yaml [hunk 1]
```diff
@@ -172,6 +194,21 @@ jobs:
...
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          role-to-assume: arn:aws:iam::${{ secrets.AWS_ACCOUNT_ID }}:role/${{ secrets.AWS_ASSUME_ROLE_NAME }}
+          role-session-name: deployInfra
+          aws-region: eu-west-2
+      - name: "Destroy Backend sandbox"
         run: |
-          npm run destroy-sandbox -- --identifier "wf-${GITHUB_RUN_ID}"
+          npm run destroy-backend-sandbox wf-${GITHUB_RUN_ID}
```
Comment (chris-elliott-nhsd) [inline .github/workflows/stage-4-acceptance.yaml:214]: you appear to have lost the teardown command for the Amplify sandbox

File: scripts/create_backend_sandbox.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  echo 1>&2 "$0: expected 1 arguments, received $#"
+  exit 2
+fi
+
+ENVIRONMENT=$1
+COMPONENT="sandbox"
+AWS_ACCOUNT_ID="$(aws sts get-caller-identity --query Account --output text)"
+AWS_REGION="eu-west-2"
+PROJECT="nhs-notify"
+GROUP="nhs-notify-template-management-nonprod"
```
Comment (chris-elliott-nhsd) [inline scripts/create_backend_sandbox.sh:16]: would this still work if we gave it a group that didn't clash with an existing group in the internal repo? seems safer to do that if possible

File: scripts/create_backend_sandbox.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  }
+}";
+
+# write the backend file
+echo -e "${backend_config}" > backend_tfscaffold.tf
+
+# clean up the file on exit
+trap "rm -f $(pwd)/backend_tfscaffold.tf" EXIT;
+
+terraform init -upgrade
```
Comment (chris-elliott-nhsd) [inline scripts/create_backend_sandbox.sh:70]: what does the extra init do?

File: .github/workflows/stage-4-acceptance.yaml (hunk 1 above)
Comment (harrim91) [reply .github/workflows/stage-4-acceptance.yaml:217]: good spot - fixed

File: scripts/create_backend_sandbox.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+terraform output -json > ${root_dir}/sandbox_tf_outputs.json
```
Comment (chris-elliott-nhsd) [inline scripts/create_backend_sandbox.sh:72]: this looks good but not 100% clear how it works. how does this know where to find the terraform outputs? is that what the backend_tfscaffold stuff is doing?

File: scripts/create_backend_sandbox.sh (hunk 3 above)
Comment (harrim91) [reply scripts/create_backend_sandbox.sh:70]: Let me try it without 😅

Comment (harrim91) [reply scripts/create_backend_sandbox.sh]: Yup, you are right. The problem was that the `terraform.sh` script creates and then deletes the backend config, so I had to manually create that again. But before I'd done that, the error I was seeing said about running `init` again. So I just added that in without really thinking it through. Thanks.

File: scripts/create_backend_sandbox.sh (hunk 2 above)
Comment (harrim91) [reply scripts/create_backend_sandbox.sh:16]: I'm not fully sure of everything `group` is for really - in this repo all I can see is it being used for tagging. In internal, it's used to pick the tfvars file. I think I added this when I was going down the route of trying to deploy this through the internal repo, and this would have been the right file to use. But I don't know if it's even needed here apart from for tagging.

File: scripts/create_backend_sandbox.sh (hunk 4 above)
Comment (harrim91) [reply scripts/create_backend_sandbox.sh:71]: Yes. The backend config block has the location of the tf state file, the `terraform output` command basically reads the state file and prints the outputs from there. There are other commands like `terraforms how` which you can use to interrogate specific resources. But this seemed like the right one to use here.

**Review by mark-r-bjss — DISMISSED** [review]
Needs platform review as well.

File: scripts/create_backend_sandbox.sh (hunk 2 above)
Comment (harrim91) [reply scripts/create_backend_sandbox.sh:16]: Maybe worth seeing what platform folk think when they review it

**Review by chris-elliott-nhsd — DISMISSED** [review]

File: .github/workflows/stage-4-acceptance.yaml [hunk 5]
```diff
@@ -60,17 +60,39 @@ jobs:
...
-      - name: "Create sandbox"
+      - name: "Create Amplify sandbox"
         run: |
-          ./scripts/create_sandbox.sh
+          ./scripts/create_amplify_sandbox.sh
       - uses: actions/upload-artifact@v4
         with:
           name: amplify_outputs.json
           path: amplify_outputs.json
+  sandbox-set-up:
```
Comment (aidenvaines-bjss) [inline .github/workflows/stage-4-acceptance.yaml:70]: I dont believe this will work. under the AMET deigns there is no ability for the component/microservice repositories to be able to do anything with the NHS Notify deployment context. The internal repo is where we manage how the services are deployed in the context of NHS Notify as the service.

@mark-r-bjss @timireland @RossBugginsNHS @m-houston and I discussed a mechanism by where there would/could be a workflow in the internal side of the programme whereby this sort of targeted creation and execution of service context testing could be run; though I don't know what happened there in my absence over the last few weeks

basically the service repos cant access AWS so this aint gunna work

**Review by aidenvaines-bjss — CHANGES_REQUESTED** [review]

File: .github/workflows/stage-4-acceptance.yaml (hunk 5 above)
Comment (aidenvaines-bjss) [reply .github/workflows/stage-4-acceptance.yaml:70]: spoke to Mark R, this is a workaround

File: infrastructure/terraform/components/sandbox/provider_aws.tf [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  ]
+
+  default_tags {
+    tags = local.default_tags
+  }
+}
+
+provider "aws" {
+  alias  = "us-east-1"
+  region = "us-east-1"
```
Comment (mark-r-bjss) [inline infrastructure/terraform/components/sandbox/provider_aws.tf:15]: Is this correct?

File: infrastructure/terraform/modules/templates-api/variables.tf [hunk 7]
```diff
@@ -55,8 +55,14 @@ variable "log_retention_in_days" {
...
   type = object({
-    USER_POOL_ID: string,
-    USER_POOL_CLIENT_ID: string
+    USER_POOL_ID : string,
+    USER_POOL_CLIENT_ID : string
   })
   description = "Cognito config"
 }
+
+variable "enable_sourcemaps" {
```
Comment (mark-r-bjss) [inline infrastructure/terraform/modules/templates-api/variables.tf:64]: I'm not sure we need a switch for this, not sure we would even switch it off.

File: infrastructure/terraform/modules/typescript-build-zip/null_resource_typescript_build.tf [hunk 8]
```diff
@@ -5,6 +5,6 @@ resource "null_resource" "typescript_build" {
 
   provisioner "local-exec" {
     working_dir = var.source_code_dir
-    command     = "npm ci && npm run build"
+    command     = "rm -rf ${var.output_dir} && npm ci && npx esbuild --bundle --minify ${var.include_sourcemaps ? "--sourcemap=inline" : ""} --target=es2020 --platform=node --outdir=${var.output_dir} ${join(" ", var.entrypoints)}"
```
Comment (mark-r-bjss) [inline infrastructure/terraform/modules/typescript-build-zip/null_resource_typescript_build.tf:8]: Would it be nicer for the TF to call a bash script to allow this to be formatted nicely and to make it easier to extend in the future?

**Review by mark-r-bjss — APPROVED** [review]

File: infrastructure/terraform/modules/templates-api/variables.tf (hunk 7 above)
Comment (harrim91) [reply infrastructure/terraform/modules/templates-api/variables.tf]: I was a bit confused about how to get it working properly - (I could only get it working with inline source maps, resulting in enormous files rather than a small file and a bigger source map) - so I thought it'd be good to have the option to turn it off. But I figured it out now, so have removed the flag.

File: infrastructure/terraform/modules/typescript-build-zip/null_resource_typescript_build.tf (hunk 8 above)
Comment (harrim91) [reply infrastructure/terraform/modules/typescript-build-zip/null_resource_typescript_build.tf]: I think that would be more complicated than it's worth now - could do it in the future if this command did get more complicated. But I can't really see it needing to get more complex than this. And putting it in a bash script adds complexity about which directory it's running from, parsing arguments etc. Whereas it's working fine like this. 😅

**Review by mark-r-bjss — APPROVED** [review]

**Review by aidenvaines-bjss — APPROVED** [review]

**Review by bhansell1 — APPROVED** [review]
